### PR TITLE
copr: Support collation aware string compare

### PR DESCRIPTION
**Pull request by breeswish**
###  What have you changed?

Based on https://github.com/tikv/tikv/pull/6562, this PR support collations in string compare built-in functions. See #6575 

###  What is the type of the changes?

- New feature (a change which adds functionality)

###  How is the PR tested?

- Unit test
- Integration test

### CODE REVIEW

File: components/tidb_query/src/expr_util/collation/utf8mb4.rs [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+pub struct CollatorUtf8Mb4GeneralCi;
+
+#[inline]
+fn general_ci_weight(c: char) -> u16 {
+    let mut plane = &GENERAL_CI_PLANE_00;
+    let r = c as u32;
+    if (r >> 8) > 0 {
+        plane = &GENERAL_CI_PLANE_01;
+    }
+    plane[(r & 0xFF) as usize]
```
Comment (andylokandy) [inline components/tidb_query/src/expr_util/collation/utf8mb4.rs]: suggestion: let plane = if (r >> 8) > 0 { &GENERAL_CI_PLANE_01 } else { &GENERAL_CI_PLANE_00 };

File: components/tidb_query/src/expr_util/collation/utf8mb4.rs [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+        let sb_chars = sb.chars();
+        Ok(sa_chars.cmp_by(sb_chars, |a, b| {
+            general_ci_weight(a).cmp(&general_ci_weight(b))
+        }))
+    }
+
+    fn sort_hash<H: Hasher>(bstr: &[u8], state: &mut H) -> Result<(), DecodeError> {
+        use std::hash::Hash;
+
+        bstr.len().hash(state);
```
Comment (andylokandy) [inline components/tidb_query/src/expr_util/collation/utf8mb4.rs]: Why is this necessary?

File: components/tidb_query/src/expr_util/collation/utf8mb4.rs (hunk 1 above)
Comment (breeswish) [reply components/tidb_query/src/expr_util/collation/utf8mb4.rs]: This part fully follows https://github.com/bb7133/proto-collations/blob/master/collations_test.go#L286 and I think keep the same is better for maintainence. Notice that actually there will be more planes in TiDB's code, so the final form should be similar to current one.

File: components/tidb_query/src/expr_util/collation/utf8mb4.rs (hunk 2 above)
Comment (breeswish) [reply components/tidb_query/src/expr_util/collation/utf8mb4.rs]: It follows Vec's hash impl.

File: components/tidb_query/src/expr_util/collation/utf8mb4.rs (hunk 1 above)
Comment (andylokandy) [reply components/tidb_query/src/expr_util/collation/utf8mb4.rs]: I think it's just the constraint in golang and we should not bring this kind of convention here. 

Further more, it's more suitable to use match when more planes is added.
